rosbag2_cpp/src/rosbag2_cpp/action_utils.cpp

```diff
@@ -0,0 +1,147 @@
+  {ActionInterfaceType::SendGoalEvent, ".+/action/.+SendGoal_Event$"},
+  {ActionInterfaceType::CancelGoalEvent, "^action_msgs/srv/CancelGoal_Event$"},
+  {ActionInterfaceType::GetResultEvent, ".+/action/.+GetResult_Event$"},
+  {ActionInterfaceType::Feedback, ".+/action/.+_FeedbackMessage$"},
+  {ActionInterfaceType::Status, "^action_msgs/msg/GoalStatusArray$"}
+};
+
+const size_t kMinActionTopicPostfixLen =
+  action_topic_to_postfix_map.at(ActionInterfaceType::Status).length();
+
+bool is_topic_belong_to_action(const std::string & topic_name, const std::string & topic_type)
+{
+  ActionInterfaceType topic = get_action_interface_type(topic_name);
+  if (topic == ActionInterfaceType::Unknown) {
+    return false;
+  }
+
+  std::regex pattern(action_topic_type_to_regex_map.at(topic));
+  return std::regex_search(topic_type, pattern);
+}
+
+std::string action_interface_name_to_action_name(const std::string & topic_name)
+{
+  std::string action_name;
+  if (topic_name.length() <= kMinActionTopicPostfixLen) {
+    return action_name;
+  } else {
+    for (const auto & [topic_type_enum, postfix] : action_topic_to_postfix_map) {
+      if (topic_name.length() > postfix.length() &&
+        topic_name.compare(topic_name.length() - postfix.length(), postfix.length(), postfix) == 0)
+      {
+        action_name = topic_name.substr(0, topic_name.length() - postfix.length());
+        break;
+      }
+    }
+  }
+
+  return action_name;
+}
+
+bool is_topic_belong_to_action(
+  ActionInterfaceType action_interface_type, const std::string & topic_type)
+{
+  if (action_interface_type == ActionInterfaceType::Unknown) {
+    return false;
+  }
+  std::regex pattern(action_topic_type_to_regex_map.at(action_interface_type));
+  return std::regex_search(topic_type, pattern);
+}
+
+std::string get_action_type(const std::string & topic_type)
+{
+  std::string action_type;
+
+  for (auto &[topic_type_enum, regex] : action_topic_type_to_regex_map) {
+    std::regex pattern(regex);
+    if (std::regex_search(topic_type, pattern)) {
+      switch (topic_type_enum) {
+        case ActionInterfaceType::SendGoalEvent:
+          // Remove the postfix "_SendGoal_Event"
+          action_type =
+            topic_type.substr(0, topic_type.length() - std::strlen("_SendGoal_Event"));
+          break;
+        case ActionInterfaceType::GetResultEvent:
+          // Remove the postfix "_GetResult_Event"
+          action_type =
+            topic_type.substr(0, topic_type.length() - std::strlen("_GetResult_Event"));
+          break;
+        case ActionInterfaceType::Feedback:
+          // Remove the postfix "_FeedbackMessage"
+          action_type =
+            topic_type.substr(0, topic_type.length() - std::strlen("_FeedbackMessage"));
+          break;
+        case ActionInterfaceType::CancelGoalEvent:
+        case ActionInterfaceType::Status:
+        default:
+          break;
```

Comment:
If we got here with some "unknown" interface type - it means that we have an error in logic somewhere. Which needs to be escalated and fixed rather than quietly swallowed. 
```suggestion
          break;
        default:
          throw std::runtime_error("Can't get action type. Unknown action interface type");
          break;
```

Reply:
Currently, we use returning an empty string to indicate that we cannot extract the action type from the action interface type. 
For example, when we scan topics to find the action name and action type. For the action type, it can be written like this (refer to: In Info::read_service_and_action_info()):
```c++
if (action_info->type.empty()) {
    action_info->type = get_action_type(topic.type);
}
```
If I use exception, The code seems to become more complex.
```c++
if (action_info->type.empty()) {
   try {
       action_info->type = get_action_type(topic.type);
   } catch (...)  {
       action_info->type = "";
   }
}
```